docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+    pub fn with_tenant<F, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub async fn with_tenant_async<F, T, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+}
+
+impl Drop for TenantAccessor {
+    // schedule a drop of itself in tokio runtime with necessary joins...
+}
+```
+
+### `TenantAccessor` and states
+
+This is how the accessor behaves when it sees different states:
+
+1. `Infant` - we cannot create a guard around the tenant at this stage. We should start loading some layers and become `Loading` or `Downloading` first and only after creating an accessor.
+2. `Loading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+3. `Broken` - we cannot run any operations on the tenant.
+4. `Downloading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+5. `Active` - the tenant can run any closures. **That means it's the only way to run any closures on tenants**.
+6. `ShuttingDown` - we cannot run any operations on the tenant.
+7. `Shutdown` - we cannot run any operations on the tenant.
+
+### `TenantAccessor` and cancelations
+
+One great question is: what guarantees do we have about tenants inside the closures we're running?
+
+After we introduced the `with_tenant` functions, there are a bunch of closures running. The only way to run a closure is when we are under an `Active` state. What if one closure wants to run under this state, and another decides to shut down the tenant? Do we want to continue executing?
+
+In this specific RFC:
+
+1. We use a `StateKeeper` / `StateSubscriber` / `StateHider` primitive to ensure all closures are gone, and running closures can't tell the difference.
+2. It's proposed to run only non-cancellable short-living closures and check in the code reviews that the closure won't run for a long time. This should be easy to spot. Moreover: _all of the current code does not use long-running functions inside tenants_, even when GC or compaction is run.
+3. To cancel GC and compaction, we could use the `CancelationSender` / `CancellationReceiver` pair. After they finish iteration, they check whether they should stop, and we wait for this to happen.
+
+## `TimelineAccessor`
+
+Works the same as `TenantAccessor`. Problems like [(#2442)](https://github.com/neondatabase/neon/issues/2442) are solved since we have to use `with_timeline` to lock on the timeline, and if GC or compaction is run, nothing will go wrong.
+
+The owner of the timeline is the tenant, as it was before.
+
+### States and transitions in `TimelineState`
+
+We can use the same states for `TimelineState`. (Do we need any additional states for `TimelineState`?)
+
+## `task_mgr` removal from tenant and timeline
```

Comment:
One important function of `task_mgr` is to enable waiting for tenant shutdown to complete. In a nutshell, the detach operation currently does this:

1. Change the tenant state from Active to Paused.
2. wait for all the tasks operating on the tenant to shut down
3. remove tenant's files from disk

With this proposal, how do you do step 2?

Reply:
Since we store `LoopedTaskController` in the tenant (previously `CancellationOwner`), we can signal GC and compaction to stop and wait until they're done. We don't need task runtime for that, only to schedule a dropping closure inside the `Drop` implementation of the tenant. We can also wait for tenant completion by subscribing to it.